File: solutions/3329-find-the-length-of-the-longest-common-prefix/solution.js
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number}
 */
// function treeNode(val){
//     this.val = val;
//     this.children = new Array(10).fill(null);
// }

var longestCommonPrefix = function(arr1, arr2) {
    const trie = new Array(10);
  
    const addToTrie = (str) => {
        let node = trie;
        for(const char of str){
            const digit = parseInt(char);
            if(node[digit] === undefined){
                node[digit] = new Array(10);
            }
            node = node[digit];
        }
    }
    
    const matchPrefix = (str) => {
        let length = 0;
        let node = trie;
        let i = 0;

        while(i < str.length){
            const digit = parseInt(str[i]);
            if(node[digit] !== undefined){
                console.log("in trie")
                length += 1;
                node = node[digit]
                i += 1;
            }else{
                break;
            }
        }
        return length;
    }
    
    for(const el of arr1){
        addToTrie(el.toString())
    };

    let max = 0;
    for(const el of arr2){
        max = Math.max(matchPrefix(el.toString()), max);
    }
        
    return max;
};

File: solutions/3329-find-the-length-of-the-longest-common-prefix/solution.py
class Solution:
    def longestCommonPrefix(self, arr1: List[int], arr2: List[int]) -> int:
        trie = {}

        for n in arr1:
            s = str(n)
            node = trie
            for c in s:
                if c not in node:
                    node[c] = {}
                node = node[c]

        best = 0
        for n in arr2:
            s = str(n)
            curr = 0
            node = trie
            for c in s:
                if c not in node:
                    break
                node = node[c]
                curr += 1
                best = max(best, curr)

        return best
